### PR TITLE
Retry failed connection

### PR DESCRIPTION
**Pull request by jeromefroe**
Hi, earlier today I posted issue [#91](https://github.com/logstash-plugins/logstash-input-jdbc/issues/91) about trying to reconnect to a database multiple times in case of a failure on the first attempt. This afternoon, I took a crack at implementing this feature myself and think I am most of the way there. I didn't add any tests, but all of the current tests still pass after my changes. Admittedly, I'm not very familiar with Ruby, but I am more than happy to go back, do some more research, and make any necessary changes or modifications. Thanks!

### CODE REVIEW

**Comment by suyograo** [comment]
@jeromefroe we'll definitely need tests. Lets us know if you need help writing tests. This blog may help you also: https://www.elastic.co/blog/logstash-functionality-through-testing

**Comment by suyograo** [comment]
Thanks for helping make this plugin better! We'll review this shortly

**Comment by jeromefroe** [comment]
@suyograo happy to help! I added two new tests to verify the plugin registers and stops without raising an exception when passed a value other than 1 (the default) for the new parameter maximum_connection_attempts. I wasn't sure what other tests to include, but I'd be glad to try and implement any suggestions.

File: lib/logstash/plugin_mixins/jdbc.rb [hunk 1]
```diff
@@ -72,6 +72,9 @@ def setup_jdbc_config
     # examples of vendor-specific options can be found in this
     # documentation page: https://github.com/jeremyevans/sequel/blob/master/doc/opening_databases.rdoc
     config :sequel_opts, :validate => :hash, :default => {}
+
+    # Maximum number of times to try connecting to database
+    config :maximum_connection_attempts, :validate => :number, :default => 1
```
Comment (suyograo) [inline lib/logstash/plugin_mixins/jdbc.rb]: Can you rename this to `connection_retry_attempts`? Also can you introduce `connection_retry_attempts_wait_time` and default to maybe `500ms`?

File: lib/logstash/plugin_mixins/jdbc.rb [hunk 2]
```diff
@@ -81,14 +84,22 @@ def jdbc_connect
...
+      retry_attempts -= 1
+      begin
+        return Sequel.connect(@jdbc_connection_string, opts=opts)
+      rescue Sequel::PoolTimeout => e
+        if retry_attempts <= 0
+          @logger.error("Failed to connect to database. #{@jdbc_pool_timeout} second timeout exceeded.")
+          raise e
+        end
+      rescue Sequel::Error => e
+        if retry_attempts <= 0
```
Comment (suyograo) [inline lib/logstash/plugin_mixins/jdbc.rb]: I think we still want to log the error and say retrying connection. And after the retry attempts are exhausted, we can raise exception

File: lib/logstash/plugin_mixins/jdbc.rb [hunk 3]
```diff
@@ -81,14 +84,22 @@ def jdbc_connect
...
+          @logger.error("Unable to connect to database", :error_message => e.message)
+          raise e
+        end
+      end
```
Comment (suyograo) [inline lib/logstash/plugin_mixins/jdbc.rb]: Introduce sleep here for `connection_retry_attempts_wait_time` ?

**Comment by jeromefroe** [comment]
@suyograo I made some changes incorporating your comments. Thanks!

**Comment by jeromefroe** [comment]
I also rebased onto master to incorporate the latest commits. That's why the old commits show up as outdated.

**Comment by dkolli** [comment]
Hi, I'm using the jdbc plug-in as well and ran into the same issue. does the retry connection make an attempt to reconnect at each interval, such as every hour in your case? this would help if the database server is down say for couple hours and up again in the third hour, logstash will try again in next interval and be able to reconnect again.

**Comment by jeromefroe** [comment]
@dkolli every time the jdbc plugin is scheduled to run, it will try to connect `connection_retry_attempts` times in the event of any connection failures, and it will wait `connection_retry_attempts_wait_time` seconds between connection attempts. The use case I had in mind was if the database was up, but the connection failed for whatever reason, you could wait briefly and try to connect again. I imagined the waiting period to be on the order of seconds, not hours though. I think using the scheduler might be better for that task.

**Comment by suyograo** [comment]
@jeromefroe can you rebase please?

**Comment by jeromefroe** [comment]
Hey @suyograo, just wanted to give you a heads up that I finished the rebase.

**Comment by untergeek** [comment]
@jeromefroe I apologize again here, but there are still conflicts because this hasn't been merged yet, and a few changes were merged since then.  Could you please rebase again?  You may need to make a few other adjustments.

**Comment by jeromefroe** [comment]
@untergeek, No need to apologize :) I rebased again, let me know if any further action is required.

**Pull request closed without merging**